services-js/registry-certs/client/birth/components/SupportingDocumentsInput.tsx

```diff
@@ -85,65 +82,79 @@ ${file.name} is ${fileSize.amount.toFixed(2) +
     this.setState({ isFocused });
   };
 
-  private updateFiles = (files: FileList): void => {
-    const { selectedFiles } = this.state;
+  private addFilesFromInput = (files: FileList): void => {
+    const { selectedFiles } = this.props;
 
     // Create an array from the FileList.
-    const fileArray: File[] = [];
+    const fileArray: UploadableFile[] = [];
 
     for (let i = 0; i < files.length; i++) {
       // Do not allow duplicate filenames, or files that are too large.
       if (
-        !selectedFiles.find(file => file.name === files[i].name) &&
+        selectedFiles &&
+        !selectedFiles.find(
+          uploadableFile => uploadableFile.file.name === files[i].name
+        ) &&
         this.checkFileSize(files[i])
       ) {
-        fileArray.push(files[i]);
-      }
-    }
+        const uploadableFile = new UploadableFile(
+          files[i],
+          this.props.uploadSessionId
+        );
 
-    this.setState(
-      { selectedFiles: [...this.state.selectedFiles, ...fileArray] },
-      () => {
-        this.props.handleChange(this.state.selectedFiles);
+        uploadableFile.upload();
 
-        // Clear actual input after each change.
-        if (this.inputRef.current) {
-          this.inputRef.current.value = null as any;
-        }
+        fileArray.push(uploadableFile);
       }
+    }
+
+    this.props.handleInputChange(
+      selectedFiles ? [...selectedFiles, ...fileArray] : fileArray
     );
+
+    // Clear native input after each change.
+    if (this.inputRef.current) {
+      this.inputRef.current.value = null as any;
+    }
   };
 
-  // Clear a file from the list.
-  private clearFile(fileToClear: File): void {
-    const filteredList = this.state.selectedFiles.filter(
-      file => file !== fileToClear
-    );
+  // Clear a file from the list, and delete from server.
+  private deleteFile = (fileToClear: File, didCancel?: boolean): void => {
+    const { selectedFiles } = this.props;
 
-    this.setState({ selectedFiles: filteredList }, () => {
-      this.props.handleChange(this.state.selectedFiles);
-    });
-  }
+    if (selectedFiles) {
+      const file = selectedFiles.find(
+        fileObject => fileObject.file === fileToClear
+      );
+
+      if (file) {
+        file.delete(didCancel).then(() => {
```

Comment:
Prefer `async`/`await` unless there’s a specific reason to go with `then`. As you saw when we worked on the `delete` method, it is easy to get mixed up over the order that the `then` block executes relative to any statements that come after it.